File: teorico/clase-18-8-2022/tads/LinkedQueue.cpp
#include "Queue.cpp"
#include <assert.h>

template <class T>
struct QueueNode
{
   T data;
   QueueNode<T> *next;
};

template <class T>
class LinkedQueue : public Queue<T>
{
private:
   int elements;
   QueueNode<T> *front;
   QueueNode<T> *back;

public:
   LinkedQueue()
   {
      elements = 0;
      front = nullptr;
      back = nullptr;
   }

   ~LinkedQueue()
   {
      while (!isEmpty())
      {
         dequeue();
      }
   }

   int size() override
   {
      return elements;
   }

   bool isEmpty() override
   {
      return elements == 0;
   }

   bool isFull() override
   {
      return false;
   }

   void enqueue(T data) override
   {
      QueueNode<T> *newNode = new QueueNode<T>;
      newNode->data = data;

      if (isEmpty())
      {
         front = newNode;
         back = newNode;
      }
      else
      {
         back->next = newNode;
         back = newNode;
      }
      elements++;
   }

   void dequeue() override
   {
      if (isEmpty())
      {
         return;
      }
      QueueNode<T> *temp = front;
      front = front->next;
      delete temp;
      elements--;
   }

   T peek() override
   {
      assert(!isEmpty());
      return front->data;
   }
};

File: teorico/clase-18-8-2022/tads/LinkedQueue.java
package tads;

public class LinkedQueue<T> implements Queue<T> {
   class Node{
      T data;
      Node next = null;
   }

   private int size;
   private Node front;
   private Node back;

   public LinkedQueue(){
      size = 0;
      front = null;
      back = null;
   }

   @Override
   public int size(){
      return size;
   }

   @Override
   public boolean isEmpty() {
      return size == 0;
   }

   @Override
   public boolean isFull(){
      return false;
   }

   @Override
   public void enqueue(T data){
      Node newNode = new Node();
      newNode.data = data;

      if(isEmpty()){
         front = newNode;
         back = newNode;
      } else{
         back.next = newNode;
         back = newNode;
      }

      size++;
   }

   @Override
   public void dequeue(){
      if(isEmpty()){
         return;
      }
      front = front.next;
   }

   @Override
   public T peek() {
      if(isEmpty()){
         throw new RuntimeException("peek on empty queue");
      }
      return front.data;
   }

}
